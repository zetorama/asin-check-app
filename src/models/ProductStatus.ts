export enum ProductStatus {
    New = 'new',
    InQueue = 'inqueue',
    Refreshing = 'refreshing',
    Success = 'success',
    Nomatch = 'nomatch',
    Timeout = 'timeout',
    Fail = 'fail',
}

export default ProductStatus
