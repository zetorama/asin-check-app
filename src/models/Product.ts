import { getParent, types, Instance, SnapshotIn } from 'mobx-state-tree'

import { StoreModel } from './Store'

export enum ProductStatus {
    Init = 'init',
    Queue = 'queue',
    Fetch = 'fetch',
    Success = 'success',
    Nomatch = 'nomatch',
    Timeout = 'timeout',
    Fail = 'fail',
}

export const ProductDetails = types.model('ProductDetails', {
    title: types.string,
    categories: types.array(types.string),
    dimensions: types.maybe(types.string),
    rank: types.maybe(types.array(types.string)),
    rating: types.maybe(types.string),
})

export type ProductAsin = string

export type ProductModel = Instance<typeof Product>
export type ProductSnapshot = SnapshotIn<typeof Product>

export const Product = types
    .model('Product', {
        asin: types.identifier,
        status: types.enumeration<ProductStatus>('ProductStatus', Object.values(ProductStatus)),
        data: types.maybe(types.frozen(ProductDetails)),

        createdOn: types.maybe(types.string),
        updatedOn: types.maybe(types.string),
        isViewed: false,
    })
    .views((self) => ({
        get isLoading(): boolean {
            return (
                self.status === ProductStatus.Init ||
                self.status === ProductStatus.Queue ||
                self.status === ProductStatus.Fetch
            )
        },
        get isCurrentlyViewing(): boolean {
            // assumming, parent has an `activeProduct` as a ref
            return getParent<StoreModel>(self, 2).viewingProduct === self
        },
    }))
    .actions((self) => ({
        toggleViewed(setAsViewed: boolean = !self.isViewed) {
            self.isViewed = setAsViewed
        },
    }))

export default Product
