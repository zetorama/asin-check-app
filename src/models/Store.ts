import { types, getParent, getSnapshot, applySnapshot, Instance, SnapshotIn } from 'mobx-state-tree'

import { fetchProductsLog, saveProductsLog, fetchProductsByAsins, refreshProductsByAsins } from '../services/product'
import { Product, ProductAsin, ProductModel, ProductStatus, ProductSnapshot } from './Product'

export const PRODUCTS_LIMIT = 12

export type StoreModel = Instance<typeof Store>
export type StoreSnapshot = SnapshotIn<typeof Store>

type Void<T> = { [P in keyof T]?: undefined }

const resetObj = <T>(source: T): Void<T> =>
    Object.keys(source).reduce((all, one) => Object.assign(all, { [one]: undefined }), {})

export const ProductInStore = types.compose(
    Product,
    types.model('ProductInStore').views((self) => ({
        get isCurrentlyViewing(): boolean {
            // assumming, parent has an `activeProduct` as a ref
            return getParent<StoreModel>(self, 2).viewingProduct === self
        },
    })),
)

export type ProductInStoreModel = Instance<typeof ProductInStore>
export type ProductInStoreSnapshot = SnapshotIn<typeof ProductInStore>

export const Store = types
    .model('AppStore', {
        version: '0.0.0',
        products: types.array(ProductInStore),
        viewingProduct: types.maybe(types.reference(ProductInStore)),
        lookingUpValue: types.optional(types.string, ''),
    })
    .views((self) => ({
        get loadingProducts(): ProductModel[] {
            return self.products.filter((p) => p.isLoading)
        },
        getProduct(asin: ProductAsin): ProductModel | undefined {
            return self.products.find((p) => p.asin === asin)
        },
    }))
    .actions((self) => ({
        mergeUpdates(products: ProductSnapshot[]) {
            // Update only existed items, keeping the order
            for (const update of products) {
                const product = self.getProduct(update.asin)
                if (product) {
                    applySnapshot(product, Object.assign(resetObj(product), update))

                    if (product === self.viewingProduct) {
                        // Mark as seen, as its currently opened
                        product.toggleViewed(true)
                    }
                }
            }
        },
        createProduct(asin: ProductAsin): ProductInStoreModel {
            const product = self.getProduct(asin)
            const defaults = product ? getSnapshot(product) : undefined
            return ProductInStore.create({
                ...defaults,
                asin,
                status: ProductStatus.New,
            })
        },
        addProducts(products: ProductModel[]) {
            if (!products || !Array.isArray(products) || !products.length) return

            const oldProducts = self.products.filter((p1) => !products.find((p2) => p1.asin === p2.asin))
            const newProducts: ProductInStoreModel[] = products.concat(oldProducts).slice(0, PRODUCTS_LIMIT)
            self.products.replace(newProducts)
        },
        clearLookingUpValue() {
            self.lookingUpValue = ''
        },
        viewProduct(product?: ProductModel | ProductAsin) {
            self.viewingProduct = product
                ? self.getProduct(ProductInStore.is(product) ? product.asin : product)
                : undefined
            if (self.viewingProduct && !self.viewingProduct.isViewed) {
                self.viewingProduct.toggleViewed(true)
            }
        },
    }))
    .actions((self) => ({
        saveProductsLog() {
            saveProductsLog(getSnapshot(self.products))
        },
        async fetchProductsLog() {
            const products = (await fetchProductsLog()) as ProductModel[]
            products && products.length && self.addProducts(products)
        },
        async lookup(value: string) {
            self.lookingUpValue = String(value || '')

            if (self.lookingUpValue) {
                const asins = self.lookingUpValue.split(/\W+/)
                const products = asins
                    .filter((asin, index) => asin && asins.indexOf(asin) === index)
                    .map((asin) => self.createProduct(asin))

                self.viewingProduct = undefined
                await self.addProducts(products)

                // switch to a newly added product
                self.viewProduct(self.products[0])
                self.clearLookingUpValue()
            }
        },
        async fetchUpdates(asins: ProductAsin[] = self.loadingProducts.map((p) => p.asin)) {
            const update = await fetchProductsByAsins(asins)
            update && self.mergeUpdates(update)
        },
        async refreshProducts(asins: ProductAsin[]) {
            for (const asin of asins) {
                const product = self.getProduct(asin)
                product && product.setStatus(ProductStatus.InQueue)
            }
            const update = await refreshProductsByAsins(asins)
            update && self.mergeUpdates(update)
        },
    }))

export default Store
