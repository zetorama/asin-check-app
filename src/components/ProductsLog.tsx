import React, { Component } from 'react'
import { ListGroup } from 'reactstrap'
import { observer } from 'mobx-react'

import ProductLogItem from './ProductLogItem'
import { StoreModel, PRODUCTS_LIMIT } from '../models/Store'
import { ProductAsin } from '../models/Product'
import autobind from '../utils/autobind'
import ListGroupItem from 'reactstrap/lib/ListGroupItem'

export interface ProductsLogProps {
    store: StoreModel
}

export class ProductsLog extends Component<ProductsLogProps> {
    render() {
        const { products, viewingProduct } = this.props.store
        if (!products || !Array.isArray(products) || !products.length) return null

        return (
            <ListGroup className='mb-3' flush>
                {products.map((product) => (
                    <ProductLogItem key={product.asin} product={product} onClick={this.handleItemClick} />
                ))}
                <ListGroupItem className='small text-muted'>
                    <em>Showing up to {PRODUCTS_LIMIT} last products</em>
                </ListGroupItem>
            </ListGroup>
        )
    }

    @autobind
    handleItemClick(e: React.MouseEvent, product: ProductAsin) {
        const { store } = this.props
        store.viewProduct(product)
    }
}

export default observer(ProductsLog)
