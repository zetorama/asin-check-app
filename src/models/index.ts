import { Store, StoreModel, StoreSnapshot } from './Store'

export * from './Store'
export * from './Product'

export function createStore(initial?: StoreSnapshot): StoreModel {
    return Store.create(initial)
}

export default createStore
