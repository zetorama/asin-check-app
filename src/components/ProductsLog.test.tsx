import * as React from 'react'
import { shallow } from 'enzyme'

import { ProductsLog } from './ProductsLog'
import { createStore, ProductStatus } from '../models'

describe('ProductsLog', () => {
    it('renders with empty store', () => {
        const store = createStore()
        const cmp = shallow(<ProductsLog store={store} />)

        expect(cmp).toMatchSnapshot()
    })

    it('renders with one product', () => {
        const store = createStore({
            products: [
                {
                    asin: 'XXX',
                    status: ProductStatus.Success,
                },
            ],
        })
        const cmp = shallow(<ProductsLog store={store} />)

        expect(cmp).toMatchSnapshot()
    })
})
