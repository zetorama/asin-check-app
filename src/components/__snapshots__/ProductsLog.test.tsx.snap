// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductsLog renders with empty store 1`] = `""`;

exports[`ProductsLog renders with one product 1`] = `
<ListGroup
  className="mb-3"
  flush={true}
  tag="ul"
>
  <ProductLogItem
    key="XXX"
    onClick={[Function]}
    product={
      Object {
        "asin": "XXX",
        "createdOn": undefined,
        "data": undefined,
        "isViewed": false,
        "status": "success",
        "updatedOn": undefined,
      }
    }
  />
  <ListGroupItem
    className="small text-muted"
    tag="li"
  >
    <em>
      Showing up to 
      12
       last products
    </em>
  </ListGroupItem>
</ListGroup>
`;
